tempesta_fw/classifier/frang.c

```diff
@@ -462,38 +462,84 @@ frang_http_ct_check(const TfwHttpReq *req, FrangAcc *ra)
 static int
 frang_http_host_check(const TfwHttpReq *req, FrangAcc *ra)
 {
+	int ret = TFW_BLOCK;
 	TfwAddr addr;
+	char *hdrhost = NULL;
+	size_t hdrhostlen = 0;
 	TfwStr field;
-	char *buf;
-	int len;
+
+	/* 1) we check, if Host: header is present.
+	 * 2) we check, that Host: header is not a IP address. 
+	 * 3) if URI has form "http://host:port/path", we check if host in URI
+	 * 	is equal to Host: header. As a particular case, if host
+	 * 	in uri is empty, Host: header also should be empty
+	 * 4) if URI has form "/path", we check, that host is not empty
+	 *
+	 * (RFC 7230 5.4):
+	 * 	A client MUST send a Host header field in all HTTP/1.1 request
+	 * 	messages.  If the target URI includes an authority component,
+	 * 	then a client MUST send a field-value for Host that is
+	 * 	identical to that authority component, excluding any userinfo
+	 * 	subcomponent and its "@" delimiter (Section 2.7.1).
+	 * 	If the authority component is missing or undefined for the
+	 * 	target URI, then a client MUST send a Host header field
+	 * 	with an empty field-value. */
+
+	TFW_DBG("Start http host check\n");
 
 	if (TFW_STR_EMPTY(&req->h_tbl->tbl[TFW_HTTP_HDR_HOST])) {
+		TFW_DBG3("Host header is missed\n");
 		frang_msg("Host header field", &ra->addr, " is missed\n");
 		return TFW_BLOCK;
 	}
-
+	TFW_DBG3("Host: header present. Length = %i, flags = 0x%04X\n",
+			 req->h_tbl->tbl[TFW_HTTP_HDR_HOST].len,
+			 (int)(req->h_tbl->tbl[TFW_HTTP_HDR_HOST].flags));
 	tfw_http_msg_hdr_val(&req->h_tbl->tbl[TFW_HTTP_HDR_HOST],
-			     TFW_HTTP_HDR_HOST, &field);
+		     TFW_HTTP_HDR_HOST, &field);
 
 	/*
-	 * FIXME: here should be a check that the Host value is not an IP
-	 * address. Need a fast routine that supports compound TfwStr.
-	 * Perhaps should implement a tiny FSM or postpone the task until we
-	 * have a good regex library.
-	 * For now just linearize the Host header field TfwStr{} string.
-	 */
-	len = field.len + 1;
-	if ((buf = tfw_pool_alloc(req->pool, len)) == NULL)
-		return TFW_BLOCK;
-	tfw_str_to_cstr(&field, buf, len);
-	if (!tfw_addr_pton(buf, &addr)) {
+	* FIXME: here should be a check that the Host value is not an IP
+	* address. Need a fast routine that supports compound TfwStr.
+	* Perhaps should implement a tiny FSM or postpone the task until we
+	* have a good regex library.
+	* For now just linearize the Host header field TfwStr{} string.
+	*/
+	hdrhostlen = field.len;
+	if ((hdrhost = tfw_pool_alloc(req->pool, hdrhostlen + 1)) == NULL) {
+		TFW_ERR("Can not allocate memory\n");
+		goto finish;
+	}
+
```

Comment:
Logically, the code above and and the code below belong to the same logical block, so I'd remove the empty line here .